### DISCUSSION TITLE
node-linker=hoisted - still get disk space benefit?

### DISCUSSION THREAD

**Discussion by GitMurf** [pnpm/pnpm#6710] [Q&A]
I have been looking long and hard for the answer to this. If you use `node-linker=hoisted` do you still get the same benefits of re-using packages and not installing the same packages over and over each new project or clone of a repo? Do you get 100% the same disk space savings? or is there some stuff that is required to be downloaded/installed/copied additionally when you need to use node-linker hoisted, and therefore you can save more disk space if you are not using node-linker=hoisted?

**Comment by gluxon** [comment]
Based on my understanding of pnpm, I would expect the answer to be yes: You would indeed still get the same disk space savings as `node-linker=isolated`.

Hard linking files from the content-addressable store is how pnpm achieves disk space savings. To confirm hard links are still used, I set up a simple project that has a dependency on `webpack` (an arbitrarily chosen dependency).

```
❯ pnpm --version
8.6.5

❯ cat .npmrc                                        
node-linker=hoisted
```

Running `pnpm install`, the output still reports the content-addressable store being used.

```
❯ pnpm install   
Lockfile is up to date, resolution step is skipped
Packages: +75
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
Packages are hard linked from the content-addressable store to the virtual store.
  Content-addressable store is at: /Users/gluxon/Library/pnpm/store/v3
  Virtual store is at:             node_modules/.pnpm
Progress: resolved 0, reused 0, downloaded 75, added 75, done
Done in 1.7s
```

We can even confirm that files in the project are indeed hard-linked using the command at https://unix.stackexchange.com/questions/201920/how-to-find-all-hard-links-to-a-given-file

```
❯ find ~/Library/pnpm/store/v3 -samefile ./node_modules/webpack/bin/webpack.js
/Users/gluxon/Library/pnpm/store/v3/files/d8/ba0c42365b1bb5bb8b58354c07ee918e3b22fcc32ca05f0ef711d84bcf43627a14d1847275162f254b6b459fe64a18ed298113fcf0abe8aa736ecd78c24d1b-exec
```
